spectator-api/src/main/java/com/netflix/spectator/api/ArrayTagSet.java

```diff
@@ -0,0 +1,252 @@
+/**
+ * Copyright 2014-2016 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spectator.api;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+/**
+ * An immutable set of tags sorted by the tag key.
+ */
+final class ArrayTagSet implements Iterable<Tag> {
+
+  private static final Comparator<Tag> TAG_COMPARATOR = (t1, t2) -> t1.key().compareTo(t2.key());
+
+  /** Empty tag set. */
+  static final ArrayTagSet EMPTY = new ArrayTagSet(new Tag[0]);
+
+  /** Create a new tag set. */
+  static ArrayTagSet create(String... tags) {
+    return EMPTY.addAll(tags);
+  }
+
+  /** Create a new tag set. */
+  static ArrayTagSet create(Tag... tags) {
+    return EMPTY.addAll(tags);
+  }
+
+  /** Create a new tag set. */
+  static ArrayTagSet create(Iterable<Tag> tags) {
+    return EMPTY.addAll(tags);
+  }
+
+  /** Create a new tag set. */
+  static ArrayTagSet create(Map<String, String> tags) {
+    return EMPTY.addAll(tags);
+  }
+
+  private final Tag[] tags;
+  private final int length;
+
+  private ArrayTagSet(Tag[] tags) {
+    this(tags, tags.length);
+  }
+
+  private ArrayTagSet(Tag[] tags, int length) {
+    if (length > tags.length) {
+      throw new IllegalArgumentException("length cannot be larger than tags array");
+    }
+    this.tags = tags;
+    this.length = length;
+  }
+
+  @Override public Iterator<Tag> iterator() {
+    return new Iterator<Tag>() {
+      private int i = 0;
+
+      @Override public boolean hasNext() {
+        return i < length;
+      }
+
+      @Override public Tag next() {
+        if (i >= length) {
+          throw new NoSuchElementException("next called after end of iterator");
+        }
+        return tags[i++];
+      }
+    };
+  }
+
+  /** Check if this set is empty. */
+  boolean isEmpty() {
+    return length == 0;
+  }
+
+  /** Add a new tag to the set. */
+  ArrayTagSet add(String k, String v) {
```

Comment:
Can be confusing on an immutable set, is there a better name? This is similar to `+` found on scala's immutable collections. This is an internal method so we can always change it later.